src/System.Net.HttpListener/src/System/Net/Managed/HttpListenerResponse.Managed.cs

```diff
@@ -142,19 +110,49 @@ public void Abort()
         private void Close(bool force)
         {
             Disposed = true;
-            _context.Connection.Close(force);
+            _httpContext.Connection.Close(force);
         }
 
         public void Close(byte[] responseEntity, bool willBlock)
         {
             CheckDisposed();
 
             if (responseEntity == null)
+            {
                 throw new ArgumentNullException(nameof(responseEntity));
+            }
 
-            ContentLength64 = responseEntity.Length;
-            OutputStream.Write(responseEntity, 0, (int)_contentLength);
-            Close(false);
+            if (_boundaryType != BoundaryType.Chunked)
+            {
+                ContentLength64 = responseEntity.Length;
+            }
+
+            if (willBlock)
+            {
+                try
+                {
+                    OutputStream.Write(responseEntity, 0, responseEntity.Length);
+                }
+                finally
+                {
+                    Close(false);
+                }
+            }
+            else
+            {
+                OutputStream.BeginWrite(responseEntity, 0, responseEntity.Length, iar =>
+                {
+                    var thisRef = (HttpListenerResponse)iar.AsyncState;
```

Comment:
I assume you do this to avoid creating a delegate, right?  Might be nice to add a comment to that effect.